web/src/main/java/org/springframework/security/web/firewall/StrictHttpFirewall.java

```diff
@@ -370,6 +474,123 @@ public FirewalledRequest getFirewalledRequest(HttpServletRequest request) throws
+					throw new RequestRejectedException("The request was rejected because the header value \"" + value + "\" is not allowed.");
+				}
+				return value;
+			}
+
+			@Override
+			public Enumeration<String> getHeaders(String name) {
+				if (!allowedHeaderNames.test(name)) {
+					throw new RequestRejectedException("The request was rejected because the header name \"" + name + "\" is not allowed.");
+				}
+				Enumeration<String> valuesEnumeration = super.getHeaders(name);
+				if (valuesEnumeration != null) {
+					while (valuesEnumeration.hasMoreElements()) {
+						String value = valuesEnumeration.nextElement();
+						if (!allowedHeaderValues.test(value)) {
+							throw new RequestRejectedException("The request was rejected because the header value \"" + value + "\" is not allowed.");
+						}
+					}
+				}
+				return super.getHeaders(name);
+			}
+
+			@Override
+			public Enumeration<String> getHeaderNames() {
+				Enumeration<String> namesEnumeration = super.getHeaderNames();
+				if (namesEnumeration != null) {
+					while (namesEnumeration.hasMoreElements()) {
+						String name = namesEnumeration.nextElement();
+						if (!allowedHeaderNames.test(name)) {
+							throw new RequestRejectedException("The request was rejected because the header name \"" + name + "\" is not allowed.");
+						}
+					}
+				}
+				return super.getHeaderNames();
+			}
+
+			@Override
+			public String getParameter(String name) {
+				if (!allowedParameterNames.test(name)) {
+					throw new RequestRejectedException("The request was rejected because the parameter name \"" + name + "\" is not allowed.");
+				}
+				String value = super.getParameter(name);
+				if (value != null && !allowedParameterValues.test(value)) {
+					throw new RequestRejectedException("The request was rejected because the parameter value \"" + value + "\" is not allowed.");
+				}
+				return value;
+			}
+
+			@Override
+			public Map<String, String[]> getParameterMap() {
+				Map<String, String[]> parameterMap = super.getParameterMap();
+				for (Map.Entry<String, String[]> entry : parameterMap.entrySet()) {
+					String name = entry.getKey();
+					String[] values = entry.getValue();
+					if (!allowedParameterNames.test(name)) {
+						throw new RequestRejectedException("The request was rejected because the parameter name \"" + name + "\" is not allowed.");
+					}
+					for (String value: values) {
+						if (!allowedParameterValues.test(value)) {
+							throw new RequestRejectedException("The request was rejected because the parameter value \"" + value + "\" is not allowed.");
+						}
+					}
+				}
+				return parameterMap;
+			}
+
+			@Override
+			public Enumeration<String> getParameterNames() {
+				Enumeration<String> namesEnumeration = super.getParameterNames();
+				while (namesEnumeration.hasMoreElements()) {
+					String name = namesEnumeration.nextElement();
+					if (!allowedParameterNames.test(name)) {
```

Comment:
I don't think we want to actually do the testing at this point. Could you return an enumeration that checks the names as they are returned?

Reply:
Certainly. Done here and in other places where an Enumeration is returned.

Lazily testing the map returned by `getParameterMap` seems quite complicated and not really worth it, so I'm still eagerly testing there. If you'd like to see a different approach there, please let me know.